File: safe_migrate_imports.py
import os
import re

package_name = "memocare"
lib_dir = os.path.abspath("lib")

def absolute_to_package(file_path, import_path):
    if import_path.startswith("package:") or import_path.startswith("dart:") or import_path.startswith("http"):
        return None
    
    # Resolve relative path
    file_dir = os.path.dirname(file_path)
    binary_path = os.path.abspath(os.path.join(file_dir, import_path))
    
    if binary_path.startswith(lib_dir):
        relative_to_lib = os.path.relpath(binary_path, lib_dir).replace("\\", "/")
        return f"package:{package_name}/{relative_to_lib}"
    return None

def process_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    def replacer(match):
        keyword = match.group(1)
        import_path = match.group(2)
        new_path = absolute_to_package(file_path, import_path)
        if new_path:
            return f"{keyword} '{new_path}'"
        return match.group(0)

    # Match import '...'; or export '...'; or part '...';
    new_content = re.sub(r"(import|export|part)\s+'([^']+)'", replacer, content)
    
    if new_content != content:
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(new_content)
        return True
    return False

# 1. Convert all imports to absolute
count = 0
for root, dirs, files in os.walk(lib_dir):
    for file in files:
        if file.endswith(".dart"):
            if process_file(os.path.join(root, file)):
                count += 1

print(f"Converted {count} files to absolute imports.")

File: safe_migrate_imports.dart
import 'dart:io';

void main() async {
  final packageName = 'memocare';
  final libDir = Directory('lib');
  final libAbsolute = libDir.absolute.path.replaceAll('\\', '/');
  final logFile = File('migration_log.txt');
  final sink = logFile.openWrite();
  int changedCount = 0;

  sink.writeln('Starting migration...');
  sink.writeln('Lib absolute: $libAbsolute');

  List<FileSystemEntity> entities = libDir.listSync(recursive: true);
  sink.writeln('Found ${entities.length} total entities in lib.');

  for (var entity in entities) {
    if (entity is File && entity.path.endsWith('.dart')) {
      final file = entity;
      final content = await file.readAsString();
      final filePath = file.absolute.path.replaceAll('\\', '/');
      final fileDir = filePath.substring(0, filePath.lastIndexOf('/'));

      final exp = RegExp(r"(import|export|part)\s+'([^']+)'");
      bool fileChanged = false;

      final newContent = content.replaceAllMapped(exp, (match) {
        final keyword = match.group(1)!;
        final importPath = match.group(2)!;

        if (importPath.startsWith('package:') ||
            importPath.startsWith('dart:') ||
            importPath.startsWith('http')) {
          return match.group(0)!;
        }

        // Manual relative resolution
        List<String> dirParts = fileDir.split('/');
        List<String> impParts = importPath.split('/');

        List<String> resolvedParts = List.from(dirParts);
        for (final part in impParts) {
          if (part == '.') continue;
          if (part == '..') {
            if (resolvedParts.isNotEmpty) resolvedParts.removeLast();
          } else {
            resolvedParts.add(part);
          }
        }

        String resolvedPath = resolvedParts.join('/');

        // Use case-insensitive check for Windows
        if (resolvedPath.toLowerCase().startsWith(libAbsolute.toLowerCase())) {
          String relativeToLib = resolvedPath.substring(libAbsolute.length);
          if (relativeToLib.startsWith('/'))
            relativeToLib = relativeToLib.substring(1);
          fileChanged = true;
          return "$keyword 'package:$packageName/$relativeToLib'";
        }

        return match.group(0)!;
      });

      if (fileChanged) {
        await file.writeAsString(newContent);
        changedCount++;
        sink.writeln('Updated: $filePath');
      }
    }
  }

  sink.writeln('FINISHED: Converted $changedCount files.');
  await sink.close();
}
